#184581172 User Sign Up

1MuhireDavid opened the pull request:
#### What does this PR do?
This PR allows enable user to be able to sign up.

#### Description of Task to be completed?
Have the following endpoint working
`POST /user/create`

#### How should this be manually tested?
- Clone this repo git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn
- Switch your working directory cd e-comm-team-amigos-bn
- Create an .env file and update it with .env-example
- Install all the dependencies by running npm install
- Then run the project by typing npm start
- Using Postman test **user/create** endpoint  
 by writing POST http://localhost:3000/user/create
_key_:Context-Type value: application/json

  

#### Any background context you want to provide?
- Have Node Js installed in your local machine
- Have Postgress and Sequelize installed on local machine
- Have Postman installed on your local machine


#### What are the relevant pivotal tracker/Trello stories?

Story ID: [ #184581172](https://www.pivotaltracker.com/n/projects/2632829/stories/184629360)

#### Screenshots
#### documented endponts on swager ui
![Screenshot (161)](https://user-images.githubusercontent.com/49757277/224570638-c6b4d511-555b-4ed7-9627-27273e06d668.png)

#### Passed Test
![Screenshot (160)](https://user-images.githubusercontent.com/49757277/224570677-b08910cb-6b15-47c1-82c8-23e1a40922cb.png)

ericndungutse committed:
chore(docker compose): setup docker for backend

 -ensure app can run in docker containers
 -create services to run express server and postgres

[Finishes #184581170]

feat(user): User signup

- resolve conflict

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(sugn up): create user account
- create  user account endpoint
- add user validation to ensure that user inputs are valid
- unit testing for the user sign up
- document user sign up endpoint with swager ui

[Delivers #1845811]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

nshutijonathan commented on .env.example:
```diff
@@ -12,3 +12,7 @@ PORT=4000
+
+# SECRET TO BE USED IN THE HASHING PASSWORD AND IN TOKENS
+ACCESS_TOKEN_SECRET=766eb85b713df63f7285459288b790c8e9e6ab43229de9dca33fe08c1c5cc30961a83456c59c9aeefafc1db3bb8d27f6964572a456fec35d2d353874409ccc78
+REFRESH_TOKEN_SECRET=a00a248ce550b46b66b08cf5ed597a0b14b1877c02714110bfe402ded54a4a3e76b426c8da9f65726c33dbe83f736ed9d905bafd4a7e942bb0e271bd499a7c6b
```
I hope these tokens are not valid ones, otherwise, we exposed them to the public.

nshutijonathan commented on src/middleware/authValidation.js:
```diff
@@ -0,0 +1,38 @@
+    await schema.validateAsync(req.body);
+    next();
+  } catch (error) {
+    console.log(error);
```
console.log is for debugging purpose ,remove it here ,It should not be pushed

nshutijonathan reviewed (changes requested):
- Please you have more than two commits and that is not acceptable 
- Can you attach the screenshot showing passing tests?
- Can you also attach a screenshot showing swagger in action /speaking of user sign up 
- Please also update your readme file and inform non-technical users how your JSON values get passed. check readme of the following repo `https://github.com/atlp-rwanda/e-comm-team-brogrammers-bn`

Habinezajanvier commented on package.json:
```diff
@@ -32,19 +32,25 @@
     "@babel/plugin-transform-runtime": "^7.21.0",
     "@babel/preset-env": "^7.20.2",
     "@babel/register": "^7.21.0",
+    "babel": "^6.23.0",
```
- Remove babel in favor to babel/cli
- Remove also body-parser in favor to express.json()

Habinezajanvier commented on package.json:
```diff
@@ -57,6 +63,7 @@
     "mocha-lcov-reporter": "^1.3.0",
     "nodemon": "^2.0.21",
     "nyc": "^15.1.0",
+    "rimraf": "^4.1.2",
```
`rimraf` is installed twice

Habinezajanvier commented on src/app.js:
```diff
@@ -1,26 +1,28 @@
+  app.use(logger("dev"));
 }
 
-app.use('/user', userRoute);
+app.use(bodyParser.json());
 
```
Remain only with `express.json()`

Habinezajanvier commented on src/controllers/authentication.js:
```diff
@@ -0,0 +1,34 @@
+
+const createUser = async (req, res) => {
+  let user = await models.User.findOne({ where: { email: req.body.email } });
+  if (user) return res.status(400).send("Email address already in use (email)");
```
Please don't send plain string as response

ericndungutse commented:
This check is repeared twice. In auth controller and in validation. **let user = await models.User.findOne({ where: { email: req.body.email } }); 
   if (user) return res.status(400).send("Email address already in use (email)");**

Emile-SHUMBUSHO commented:
> This check is repeared twice. In auth controller and in validation. **let user = await models.User.findOne({ where: { email: req.body.email } });   if (user) return res.status(400).send("Email address already in use (email)");**

nyakamwe committed:
ft(signUp):sign up a user
- creates signUp endpoint
- validate user inputs
- validate user inputs if are valid
- generate jwt on successfull user registration
- create a unut testing for signUp
- documents user signUp endpoint with swager ui

[Delivers #184581172]

Habinezajanvier reviewed (approved):
👍
